util/argo/argo.go

```diff
@@ -1002,33 +944,43 @@ func GetPermittedRepos(proj *argoappv1.AppProject, repos []*argoappv1.Repository
 	return permittedRepos, nil
 }
 
-func getDestinationServer(ctx context.Context, db db.ArgoDB, clusterName string) (string, error) {
-	servers, err := db.GetClusterServersByName(ctx, clusterName)
-	if err != nil {
-		return "", fmt.Errorf("error getting cluster server by name %q: %w", clusterName, err)
-	}
-	if len(servers) > 1 {
-		return "", fmt.Errorf("there are %d clusters with the same name: %v", len(servers), servers)
-	} else if len(servers) == 0 {
-		return "", fmt.Errorf("there are no clusters with this name: %s", clusterName)
-	}
-	return servers[0], nil
+type ClusterGetter interface {
+	GetCluster(ctx context.Context, name string) (*argoappv1.Cluster, error)
+	GetClusterServersByName(ctx context.Context, server string) ([]string, error)
 }
 
-func getDestinationServerName(ctx context.Context, db db.ArgoDB, server string) (string, error) {
-	if db == nil {
-		return "", errors.New("there are no clusters registered in the database")
-	}
-
-	cluster, err := db.GetCluster(ctx, server)
-	if err != nil {
-		return "", fmt.Errorf("error getting cluster name by server %q: %w", server, err)
+// GetDestinationCluster returns the cluster object based on the destination server or name. If both are provided or
+// both are empty, an error is returned. If the destination server is provided, the cluster is fetched by the server
+// URL. If the destination name is provided, the cluster is fetched by the name. If multiple clusters have the specified
+// name, an error is returned.
+func GetDestinationCluster(ctx context.Context, destination argoappv1.ApplicationDestination, db ClusterGetter) (*argoappv1.Cluster, error) {
+	if destination.Name != "" && destination.Server != "" {
+		return nil, fmt.Errorf("application destination can't have both name and server defined: %s %s", destination.Name, destination.Server)
 	}
```

Comment:
Is this constraint existing at the CRD level today? I suspect that nothing fails today if both fields are present. Can you please confirm?

Reply:
Not exactly at the CRD level, but it is currently enforced at runtime: https://github.com/argoproj/argo-cd/pull/21189/files#diff-87b3b573e8c40af95af4dbda83f7dbc467219b02182415ff8e1f0ad73a82dc59L518

I tested locally in 2.14.0-rc5. If I set both fields, the app goes into an Unknown state with an error `application destination can't have both name and server defined`.